src/test/groovy/graphql/nadel/engine/NadelExecutionStrategyTest2.groovy

```diff
@@ -1062,6 +1062,239 @@ class NadelExecutionStrategyTest2 extends StrategyTestHelper {
+                "UserService",
+                userServiceSchema,
+                query,
+                ["issues"],
+                expectedQuery1,
+                [issues: [issue1, issue2]],
+                expectedQuery2,
+                [usersByIds: [user1, null, user2]]
+        )
+
+
+        then:
+        def user1Result = [name: 'Name']
+        def user2Result = [name: 'Name 2']
+        def issue1Result = [id: "ISSUE-1", authors: [user1Result]]
+        def issue2Result = [id: "ISSUE-2", authors: [null, user2Result]]
+        response == [issues: [issue1Result, issue2Result]]
+        errors.size() == 0
+    }
+
+    def "hydration matching using index with arrays"() {
+        given:
+        def issueSchema = TestUtil.schema("""
+        type Query {
+            issues : [Issue]
+        }
+        type Issue {
+            id: ID
+        }
+        """)
+        def userServiceSchema = TestUtil.schema("""
+        type Query {
+            usersByIssueIds(issueIds: [ID]): [[User]]
+        }
+        type User {
+            id: ID
+            name: String
+        }
+        """)
+
+        def overallSchema = TestUtil.schemaFromNdsl('''
+        service Issues {
+            type Query {
+                issues: [Issue]
+            }
+            type Issue {
+                id: ID
+                authors: [User] => hydrated from UserService.usersByIssueIds(issueIds: $source.id) object indexed, batch size 5
+            }
+        }
+        service UserService {
+            type Query {
+                usersByIssueIds(issueIds: [ID]): [[User]]
+            }
+            type User {
+                id: ID
+                name: String
+            }
+        }
+        ''')
+
+        def query = '{issues {id authors {name} }}'
+        def expectedQuery1 = "query nadel_2_Issues {issues {id id}}"
+        def issue1 = [id: "ISSUE-1"]
+        def issue2 = [id: "ISSUE-2"]
+
+        def expectedQuery2 = "query nadel_2_UserService {usersByIssueIds(issueIds:[\"ISSUE-1\",\"ISSUE-2\"]) {name}}"
+        def user1 = [name: 'Name']
+        def user2 = [name: 'Name 2']
+
+        Map response
+        List<GraphQLError> errors
+        when:
+        (response, errors) = test2Services(
+                overallSchema,
+                "Issues",
+                issueSchema,
+                "UserService",
+                userServiceSchema,
+                query,
+                ["issues"],
+                expectedQuery1,
+                [issues: [issue1, issue2]],
+                expectedQuery2,
+                [usersByIssueIds: [[user1], [user1, user2]]]
+        )
+
+
+        then:
+        def user1Result = [name: 'Name']
+        def user2Result = [name: 'Name 2']
+        def issue1Result = [id: "ISSUE-1", authors: [user1Result]]
+        def issue2Result = [id: "ISSUE-2", authors: [user1Result, user2Result]]
+        response == [issues: [issue1Result, issue2Result]]
+        errors.size() == 0
+    }
+
```

Comment:
I think we should have a test where the contract is NOT upheld and hence we get a runtime assertion

This is really important I think (I mean it will index out of bounds exception anyway but still)

Reply:
Can do.

- [ ] Test case for contract not valid